### ISSUE TITLE
topics with id of 0

### ISSUE THREAD

**Issue by drfraser** [therealexpertai/nlapi-python#2]
Hi,

I am using version 2.10 and Python 3.8

I was trying out your service and can not get anything to work using either full_analysis() or specific_resource_analysis()

The problem seems to be topics with ids of 0.  In making an API call, the data does get to the server and is processed (I can see my character usage go up), but the response is not being processed properly. Here is the relevant section of the stack trace:

  File "/Volumes/Phil/projects/research/holistic/src/holistic/features/featset/expertai.py", line 157, in make_features
    results = self.client.specific_resource_analysis(
  File "/Volumes/Phil/projects/research/holistic/venv/lib/python3.8/site-packages/expertai/nlapi/cloud/client.py", line 101, in specific_resource_analysis
    return self.process_response(response)
  File "/Volumes/Phil/projects/research/holistic/venv/lib/python3.8/site-packages/expertai/nlapi/cloud/client.py", line 79, in process_response
    return ObjectMapper().read_json(response.json)
  File "/Volumes/Phil/projects/research/holistic/venv/lib/python3.8/site-packages/expertai/nlapi/cloud/object_mapper.py", line 106, in read_json
    dm = DataModel(**data)
  File "/Volumes/Phil/projects/research/holistic/venv/lib/python3.8/site-packages/expertai/nlapi/common/model/data_model.py", line 88, in __init__
    self._topics = [Topic(**t) for t in topics]
  File "/Volumes/Phil/projects/research/holistic/venv/lib/python3.8/site-packages/expertai/nlapi/common/model/data_model.py", line 88, in <listcomp>
    self._topics = [Topic(**t) for t in topics]
  File "/Volumes/Phil/projects/research/holistic/venv/lib/python3.8/site-packages/expertai/nlapi/common/model/topic.py", line 37, in __init__
    raise MissingArgumentError("Missing required argument: id")
expertai.nlapi.common.errors.MissingArgumentError: Missing required argument: id

I patched client.py to print out the raw response from the server, here is the relevant section (all topics do have id  numbers):

    "topics": [
      {
        "id": 114,
        "label": "construction industry",
        "score": 4.09,
        "winner": true
      },
      {
        "id": 0,
        "label": "clothing",
        "score": 0.6,
        "winner": false
      },

Looking at topic.py, the problem is obviously

        if not (id or id_):
            raise MissingArgumentError("Missing required argument: id")

Testing "if not (0 or None)", the condition succeeds.  So to get things working I have changed it to 

        if id is None and id_ is None:
            raise MissingArgumentError("Missing required argument: id")

The line assigning a value to self._id also needs fixed, of course.  I don't know if ids can actually be 0s (you tell me, probably not but they are being returned), but to get further done down the road, I will assume so for the time being.

Otherwise, the return data I see from the test looks ok, but I need to check the details.  Thanks for the free account, I'm doing some MSc research.

P.S. I doubt I'll hit 10 million characters in a month, but if I do, any chance of a credit/reset on the count? - this is a pretty basic bug I wouldn't have expected to run into, so I am really wondering what happened.  email is dfraser@douglas-fraser.com

Doug

**Comment by avarone-github** [comment]
Good catch Doug, thank you, the issue you found is going to be fixed soon adopting your suggestion.

The consumed characters counter is reset at the beginning of each calendar month, so, unless you are going to submit 10 million characters before December 1st, you should have no problem.

**Comment by marcobellei-eai** [comment]
Hi Doug,

we just commited your fix and deployed v2.1.1 on pypi.
Please, let us know if the problem is solved.

Thanks
Marco

**Comment by drfraser** [comment]
Hi Marco,

I've downloaded v2.1.1 and looking at the code, I see the line in topic.py actually assigning self._id did not change.  I mentioned that problem, but didn't specify what I had patched as I felt you should do what you felt was right (coding style etc). i.e.

        self._id = id or id_

will break if id is 0 and id_ is None because "0 or None" = None (not 0).  I saw in a help page that topics can have ids of 0 so that's ok then. I think this line should be changed to (at least something like):

    self._id = id if id is not None else id_

That is not the greatest line of code, but it deals with the problem. I can't think of how to otherwise; the problem you are trying to address with having both 'id' and 'id_'  I understand from the comment, but it seems a little out of the ordinary (I don't fully understand the larger context, TBH).

Doug

**Comment by marcobellei-eai** [comment]
Hi Doug,

we just commited the id fix you suggested and deployed v2.1.2 on pypi.
Please, let us know if the problem is solved.

Thanks
Marco

**Comment by drfraser** [comment]
Hi,

I uninstalled 2.1.0 and installed 2.1.2.  Looking at topic.py, the new lines look like what I'd expect.  But running my code, I get "Failed to fetch the Bearer Token. Error: 500-Internal Server Error" as a response from your servers.

If I uninstall 2.1.2 and immediately install 2.1.0, then things work.  Reinstalling 2.1.2 results in the same error again.

I guess I'd say close this issue because the fix in 2.1.2 should work.  The major problem now lies with the other issue I have opened.  I can't imagine it lies on my side because reinstalling 2.1.0 clears things up.
